### DISCUSSION TITLE
Change filter of graphql object listing query before loading the actual objects

### DISCUSSION THREAD

**Discussion by kursph** [pimcore/pimcore#15753] [Q&A / Support] [answered]
Hello.

Using data-hub and the endpoint for listing objects, I wanted to ask whether or not I can fetch and change the filter value within the Schema before the actual listing gets loaded in `Pimcore\Bundle\DataHubBundle\Event\GraphQL\Model\ListingEvent` when calling upon `Pimcore\Bundle\DataHubBundle\Event\GraphQL\ListingEvents::PRE_LOAD`?

Reason being is that I want to avoid loading thousands of, _Products_ in this case, objects just to, after the actual list is already loaded, filter the many objects just to return a specific few --> hence I want to keep the query parameterless like this one i.e `{
  getProductListing {
    edges {
      node {
        matNr
      }
    }
  }
}`

**Comment by kursph** [comment]
Some more context --> in the FE the parameterless query should be called because the JWT returned from the Azure API Gateway includes a header with an id that is required as reference for listing the correct objects. So before actually loading the objects I'd like to take the header from the request and filter the objects accordingly.

**Comment by SamyMP** [comment]
Hello, 

I don't know if I understood your question very well, but you can use filters in your query, like this : 
```graphql
{
  getProductListing(
    filter: "{\"someStringProperty\": {\"$like\" :\"whatever%\"}, \"numberProperty\": 1}"
  ) {
    edge {
      node {
        matNr
      }
    }
  }
}
```

Here is the documentation : https://pimcore.com/docs/platform/Datahub/GraphQL/Query/Filtering

**Reply by kursph** [reply]
thanks, but I want to define the query without any parameters like this `getProductListing { query body }` and set a filter afterwards during e.g the `Pimcore\Bundle\DataHubBundle\Event\GraphQL\ExecutorEvents::PRE_EXECUTE` event.

**Reply by SamyMP** [reply]
In that case, you're probably looking for this doc : https://pimcore.com/docs/platform/Datahub/GraphQL/Events#example-4-add-custom-query-conditions-to-object-listing

😉

**Reply by kursph** [reply]
yes I already checked it out, but not quite what I was looking for. Because here -->
```
if ($listing->getClassName() === 'Product') {
          $listing->setCondition(sprintf('(%s IN (SELECT %s FROM objects WHERE %s=\'object\') AND %s = \'variant\')',
          Service::getVersionDependentDatabaseColumnName('o_parentid'),
          Service::getVersionDependentDatabaseColumnName('o_id'),
          Service::getVersionDependentDatabaseColumnName('o_type'),
          Service::getVersionDependentDatabaseColumnName('o_type')));
        }
```

before `$listing->setCondition( ... )` data from the dataObjects is already being loaded into the listing which is precisely what I want to avoid. Therefor, before any data is being loaded into the listing, that's where I want to set a filter.

**Reply by SamyMP** [reply, marked as answer]
> before $listing->setCondition( ... ) data from the dataObjects is already being loaded into the listing

Actually, I don't think so. The way listing works is : as long as you don't call the `load()` or `getData()` methods, nothing is loaded into memory. That's why the event is called `ListingEvents::PRE_LOAD`

I confirmed that by trying this EventListener, and putting a breakpoint in there :
```php
/**
     * @param ListingEvent $event
     */
    public function onListingPreLoad(ListingEvent $event)
    {
        $listing = $event->getListing();

        if ($listing->getClassName() === 'Product') {  // <- breakpoint here
          $listing->setCondition(sprintf('(%s IN (SELECT %s FROM objects WHERE %s=\'object\') AND %s = \'variant\')',
          Service::getVersionDependentDatabaseColumnName('o_parentid'),
          Service::getVersionDependentDatabaseColumnName('o_id'),
          Service::getVersionDependentDatabaseColumnName('o_type'),
          Service::getVersionDependentDatabaseColumnName('o_type')));
        }

        $event->setListing($listing);
    }
```

At this moment, printing `$listing` in the console resultated in "null" for the data property. However, calling `getData()` actually loaded the listing with the given condition/filters from my query in the playground.

So I think you can use this event, it matches your needs 😊
